lib/internal/promise_hooks.js

```diff
@@ -0,0 +1,121 @@
+'use strict';
+
+const {
+  ArrayPrototypeIndexOf,
+  ArrayPrototypeSlice,
+  ArrayPrototypeSplice,
+  ArrayPrototypePush,
+  FunctionPrototypeBind
+} = primordials;
+
+const { setPromiseHooks } = internalBinding('async_wrap');
+const { triggerUncaughtException } = internalBinding('errors');
+
+const hooks = {
+  init: [],
+  before: [],
+  after: [],
+  resolve: []
+};
+
+function initAll(promise, parent) {
+  const hookSet = ArrayPrototypeSlice(hooks.init);
+  const exceptions = [];
+
+  for (let i = 0; i < hookSet.length; i++) {
+    const init = hookSet[i];
+    try {
+      init(promise, parent);
+    } catch (err) {
+      ArrayPrototypePush(exceptions, err);
+    }
+  }
+
+  // Triggering exceptions is deferred to allow other hooks to complete
+  for (let i = 0; i < exceptions.length; i++) {
+    const err = exceptions[i];
+    triggerUncaughtException(err, false);
+  }
+}
+
+function makeRunHook(list) {
+  return (promise) => {
+    const hookSet = ArrayPrototypeSlice(list);
+    const exceptions = [];
+
+    for (let i = 0; i < hookSet.length; i++) {
+      const hook = hookSet[i];
+      try {
+        hook(promise);
+      } catch (err) {
+        ArrayPrototypePush(exceptions, err);
+      }
+    }
+
+    // Triggering exceptions is deferred to allow other hooks to complete
+    for (let i = 0; i < exceptions.length; i++) {
+      const err = exceptions[i];
+      triggerUncaughtException(err, false);
+    }
+  };
+}
+
+const beforeAll = makeRunHook(hooks.before);
+const afterAll = makeRunHook(hooks.after);
+const resolveAll = makeRunHook(hooks.resolve);
+
+function maybeFastPath(list, runAll) {
+  return list.length > 1 ? runAll : list[0];
+}
+
+function update() {
+  const init = maybeFastPath(hooks.init, initAll);
+  const before = maybeFastPath(hooks.before, beforeAll);
+  const after = maybeFastPath(hooks.after, afterAll);
+  const resolve = maybeFastPath(hooks.resolve, resolveAll);
+  setPromiseHooks(init, before, after, resolve);
+}
+
+function stop(list, hook) {
+  const index = ArrayPrototypeIndexOf(list, hook);
+  if (index >= 0) {
+    ArrayPrototypeSplice(list, index, 1);
+    update();
+  }
+}
+
+function makeUseHook(list) {
+  return (hook) => {
+    ArrayPrototypePush(list, hook);
+    update();
+    return FunctionPrototypeBind(stop, null, list, hook);
+  };
+}
+
+const onInit = makeUseHook(hooks.init);
+const onBefore = makeUseHook(hooks.before);
+const onAfter = makeUseHook(hooks.after);
+const onResolve = makeUseHook(hooks.resolve);
+
+function createHook({ init, before, after, resolve } = {}) {
+  const hooks = [];
+
+  if (init) ArrayPrototypePush(hooks, onInit(init));
```

Comment:
It might make sense to document that the hooks are called on a first-come-first-served basis, and test this behavior (in conjunction with async hooks) - describing it as being similar to async hooks probably implies that already, but it would be better to be explicit (and save the users some time digging into another doc with a sentence)

Reply:
I'm not sure I understand what you mean by that. Are you referring to that promise hooks are at an earlier than async hooks? Or are you referring to the behaviour when registering multiple hook sets?

In a case where you have both a promise hook set and a separate async hooks set, they could occur in either order depending on which set was initialized first. I'm not sure it really makes sense to imply any sort of determinism in the ordering of that given the timing of the event forwarding from promise hooks to async hooks being immediate and therefore possibly having promise hook triggered async hooks trigger before another set of promise hooks.

Reply:
I agree it's fine to leave it undefined, though in that case we probably want to explicitly document that it is undefined, too (just in case that the user depends on the whatever the behavior they run into).

Reply:
I agree it's fine to leave it undefined, though in that case we probably want to explicitly document that it is undefined, too (just in case that the user depends on the whatever the behavior they run into).